### PR TITLE
Node demo application

### PR DESCRIPTION
**Pull request by esonmez**
## Description
Basic demo application in Node using our [node SDK](https://github.com/smartcar/node-sdk), express, handlebars, and bootstrap.

This demo is paired with our [“Running our demo app”](https://support.smartcar.com/getting-started/run-our-demo-app) guide. 

### Context
The goal of our demo applications is to make it quick and easy to get started with Smartcar. 

Specifically:
- Direct the user to [configure an app on Smartcar](https://support.smartcar.com/getting-started/configure-your-application).
- Allow the user to authenticate with a vehicle using their project configuration. The user should be able to launch our connect flow and exchange an authorization code for an access token.
- Allow the user to use this access token to make requests to the vehicle.
- Open source our demo applications so users can clone the repo, run the app locally, and view the source code to learn more about how Smartcar works.

Out of scope:
- Refresh token - this version of our demo app focuses purely on making a request using an access token.
- Javascript SDK - I used the Node SDK only to keep the example simple.
- Unit tests - I left unit tests out to keep the repo light & simple. These can be added at a later date if needed.

## Test Plan
_Basic flow:_
1. [Configure an application](https://developer.smartcar.com) on Smartcar.
2. Clone this branch locally:
```bash
git clone git@github.com:smartcar/demo.git
git checkout node-demo
cd demo/node
npm install
```
3. Run the application.
```bash
# For security, you must configure http://localhost:8000/callback as a redirect uri in Smartcar's developer portal.
PORT=8000 \
SMARTCAR_CLIENT_ID=[CLIENT_ID] \
SMARTCAR_SECRET=[CLIENT_SECRET] \
SMARTCAR_REDIRECT_URI=http://localhost:8000/callback \
SMARTCAR_MODE=development \
node app.js
# Navigate to http://localhost:8000
```
4. Click "Connect your car." Confirm the app redirects to connect, then redirects back on successful authentication.
5. Select a vehicle. For each request type (info, location, odometer, lock, unlock), click "Send request." Confirm the request succeeds and that the data shows at the `/request` route.
6. On the `/request` route, click "Start over." Confirm the page navigates to step 1. Connect the car again and, on Step 2, click "Start over." Confirm the page navigates back to step 1.

_Additional tests:_
- When connecting the vehicle, fill out the "Don't see your brand" form. Confirm connect redirects to Step 1 of the demo.
- To test error handling of environment variables, run the app with incorrect or missing environment variables,

## Next steps
- [x] Open source this repo.
- [x] Update our [getting started guide](https://support.smartcar.com/getting-started/run-our-demo-app) to pair the guide with this demo.

### CODE REVIEW

File: node/app.js [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+      const vehicles = {};
+      _.forEach(vehicleIds, vehicleId => {
+        vehicles[vehicleId] = {
+          id: vehicleId,
+          instance: new smartcar.Vehicle(vehicleId, ACCESS_TOKEN),
+        };
+      });
+
+      // Add vehicle info to vehicle objects
+      const vehicleInfoPromises = _.map(vehicles, ({instance}) => instance.info());
```
Comment (GregChan) [inline node/app.js]: Rename instance to vehicle.

File: node/app.js [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+});
+
+app.get('/vehicles', function(request, response, next) {
+
+  if (!ACCESS_TOKEN) {
+    return response.redirect('/');
+  }
+
+  smartcar.getVehicleIds(ACCESS_TOKEN)
```
Comment (sydneypalumbo) [inline node/app.js]: You can shorten lines 105 - 127 by taking out one of the ``_.forEach``
I did something like this in the app I'm working on (if you initialize VEHICLES as ``VEHICLES={}`` this works in your demo too just as is)
```    
smartcar.getVehicleIds(ACCESS_TOKEN)
        .then((res)=>{
            const vehicleIds = res.vehicles
            const vehiclePromises = vehicleIds.map((vehicleId) => {
                const vehicle = new smartcar.Vehicle(vehicleId, ACCESS_TOKEN)
                VEHICLES[vehicleId] = {id: vehicleId, instance: vehicle}
                return vehicle.info()
            });
            Promise.all(vehiclePromises)
                .then((vehicles)=> {
                    response.render('vehicles', {vehicles, response: RESPONSE});
                })
        })
```

File: node/app.js [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+// Setting MODE to "development" will show Smartcar's mock vehicle
+const SMARTCAR_MODE = envvar.oneOf('SMARTCAR_MODE', ['development', 'production'], 'development');
+
+// For the purposes of this demo, we store the access token in memory.
+// For a production application, please store the access token in a persistent
+// data store.
+let ACCESS_TOKEN = null;
+
+let VEHICLES = null;
+let RESPONSE = null;
```
[the remark this answers is not in the document]
Comment (esonmez) [reply node/app.js]: Removed

File: node/app.js (hunk 2 above)
Comment (esonmez) [reply node/app.js]: Awesome. I updated to this version.

One small adjustment - I use the vehicle info in the `/data` route to render which vehicle the developer sent the request to.  I added a `_.forEach` in the second block to update `VEHICLES` with the vehicle info. Let me know if we can clean this up further :fist_oncoming: 

```javascript
      return Promise.all(vehiclePromises)
        .then(function(vehicles) {

          // ADDED
          // Add vehicle info to vehicle objects
          _.forEach(vehicles, vehicle => {
            const {id: vehicleId} = vehicle;
            VEHICLES[vehicleId] = Object.assign(VEHICLES[vehicleId], vehicle);
          });

          res.render('vehicles', {vehicles});
        })
        .catch(function(err) {
          const message = err.message || 'Failed to get vehicle info.';
          const action = 'fetching vehicle info';
          return redirectToError(res, message, action);
        });
```

File: node/app.js [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+// Use handlebars
+app.engine('.hbs', exphbs({
+  defaultLayout: 'main',
+  extname: '.hbs',
+}));
+app.set('view engine', '.hbs');
+
+// Routes
+app.get('/', function(request, response, next) {
```
[the remark this answers is not in the document]
Comment (esonmez) [reply node/app.js]: Updated to `req` and `res`

**Comment by esonmez** [comment]
@GregChan @sydneypalumbo I made a series of updates. A few notes...

1. The `/request` method is still quite long -- let me know if you think we should shorten it further or if it is clear enough as is.

2. See my note above: re adding the vehicle info to each vehicle object within VEHICLES.

3. Thoughts on A vs B for the vehicle object schema within VEHICLES? The second makes sense given our naming in our Node SDK. The first seems to make more sense given the schema (`VEHICLES[vehicleId].vehicle` seems a little funky).

```javascript
// Option A

VEHICLES[vehicleId] = {
  id: 'VEHICLE_ID',
  instance: {}, // vehicle instance
  make: 'VEHICLE_MAKE'
  model: 'VEHICLE_MODEL',
  year: 'VEHICLE_YEAR',
}

//Option B

VEHICLES[vehicleId] = {
  id: 'VEHICLE_ID',
  vehicle: {}, // vehicle instance
  make: 'VEHICLE_MAKE'
  model: 'VEHICLE_MODEL',
  year: 'VEHICLE_YEAR',
}

```

File: node/app.js [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+          const action = 'fetching vehicle info';
+          return redirectToError(res, message, action);
+        });
+      break;
+    case 'location':
+      instance.location()
+        .then(({data}) => res.render('data', {data, type, vehicle}))
+        .catch(function(err) {
+          const message = err.message || 'Failed to get vehicle location.';
+          const action = 'fetching vehicle location';
```
Comment (GregChan) [inline node/app.js]: Can we change the `action` to just the endpoint name? You could probably template out the error message and the action in the `redirectToError` function, if you want to cut down the number of lines. I'm not sure if that would make it easier to read.. @sydneypalumbo thoughts?

File: node/package.json [hunk 6]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (sankethkatta) [inline node/package.json]: @esonmez re: your friend's email on needing to use node 6. Using the "engines" field here should solve that: https://docs.npmjs.com/files/package.json#engines

File: node/app.js (hunk 5 above)
Comment (esonmez) [reply node/app.js]: @gregchan I use redirectToError in other places in app.js that aren't requests to data endpoints. I could add another helper that takes in an endpoint, templates the action, and calls `redirectToError` but I think that may be overkill.

Comment (GregChan) [reply node/app.js]: Got it

Comment (GregChan) [reply node/app.js]: If we're optimizing for code readability, I'd argue that having less useful errors and reducing the amount of error handling code would be better. If we're optimizing for user experience of the demo app, then these error messages are useful. 

I'm of the opinion we should reduce the number of lines for readability, but I'm not sure how to do this without sacrificing error message quality. 

It's your call.

**Comment by esonmez** [comment]
@GregChan I agree. I'll keep error messages as is for the UX of the app :rocket:

**Pull request merged**